### ISSUE TITLE
about dropoutLayer

### ISSUE THREAD

**Issue by zhongtao93** [tbennun/cudnn-training#4]
I find that there is no dropout function in CUDnn, and from your code I think it's hard to build a new dropout function with CUDA by myself. Do you have any experience in build a dropoutLayer with CUDA?

**Comment by tbennun** [comment]
A dropout layer is pretty easy to implement. In essence, forward prop zeros out random elements of the previous layer. Keeping the elements you zeroed, the backprop phase propagates back the derivatives from the nonzero elements and 0s from the others. See Caffe's implementation:
https://github.com/BVLC/caffe/blob/master/src/caffe/layers/dropout_layer.cu#L9

In the training example, you would first have to include a random state in the TrainingContext object (I would use [cuRAND](http://docs.nvidia.com/cuda/curand/)), allocate an empty buffer for the random values, and with each SGD step randomize that buffer using cuRAND.

The forward and backprop kernels can be implemented and called as the example shows with Softmax:
https://github.com/tbennun/cudnn-training/blob/master/lenet.cu#L266

Good luck

**Comment by zhongtao93** [comment]
Ok, Thank you very much for your detail explanation, I'll go to implement it! Thank you !
